File: src/main/java/com/luneo7/junit4/intellij/maven/Importer.java
package com.luneo7.junit4.intellij.maven;

import com.intellij.openapi.externalSystem.service.project.IdeModifiableModelsProvider;
import com.intellij.openapi.module.Module;
import com.luneo7.junit4.intellij.state.ListenersConfig;
import org.jdom.Element;
import org.jetbrains.idea.maven.importing.MavenImporter;
import org.jetbrains.idea.maven.importing.MavenRootModelAdapter;
import org.jetbrains.idea.maven.project.MavenProject;
import org.jetbrains.idea.maven.project.MavenProjectChanges;
import org.jetbrains.idea.maven.project.MavenProjectsProcessorTask;
import org.jetbrains.idea.maven.project.MavenProjectsTree;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

public class Importer extends MavenImporter {

    public Importer() {
        super("org.apache.maven.plugins", "maven-surefire-plugin");
    }

    @Override
    public void preProcess(Module module,
                           MavenProject mavenProject,
                           MavenProjectChanges changes,
                           IdeModifiableModelsProvider modifiableModelsProvider) {
        // no-op
    }

    @Override
    public void process(IdeModifiableModelsProvider modifiableModelsProvider,
                        Module module,
                        MavenRootModelAdapter rootModel,
                        MavenProjectsTree mavenModel,
                        MavenProject mavenProject,
                        MavenProjectChanges changes,
                        Map<MavenProject, String> mavenProjectToModuleName,
                        List<MavenProjectsProcessorTask> postTasks) {
        if (module == null) {
            return;
        }

        ListenersConfig service = module.getService(ListenersConfig.class);

        if (mavenProject != null) {
            List<String> listeners = mavenProject.getPlugins()
                                                 .stream()
                                                 .filter(m -> "maven-surefire-plugin".equals(m.getArtifactId()))
                                                 .findFirst()
                                                 .flatMap(mavenPlugin -> {
                                                     if (mavenPlugin.getConfigurationElement() != null && mavenPlugin.getConfigurationElement()
                                                                                                                     .getChildren() != null) {
                                                         return mavenPlugin.getConfigurationElement()
                                                                           .getChildren()
                                                                           .stream()
                                                                           .filter(e -> "properties".equals(e.getName()))
                                                                           .flatMap(e -> e.getChildren()
                                                                                          .stream())
                                                                           .filter(e -> e.getChildren()
                                                                                         .stream()
                                                                                         .anyMatch(e1 -> "name".equals(e1.getName()) &&
                                                                                                 "listener".equals(e1.getValue())))
                                                                           .flatMap(e -> e.getChildren()
                                                                                          .stream()
                                                                                          .filter(e1 -> "value".equals(e1.getName()))
                                                                                          .map(Element::getValue))
                                                                           .findFirst();


                                                     }
                                                     return Optional.empty();
                                                 })
                                                 .map(l -> Arrays.stream(l.split(","))
                                                                 .map(String::trim)
                                                                 .filter(e -> !e.isEmpty())
                                                                 .collect(Collectors.toList()))
                                                 .orElse(Collections.emptyList());

            service.getState().setListeners(listeners);
        }
    }
}

File: src/main/kotlin/com/luneo7/junit4/intellij/maven/Importer.kt
package com.luneo7.junit4.intellij.maven

import com.intellij.openapi.externalSystem.service.project.IdeModifiableModelsProvider
import com.intellij.openapi.module.Module
import org.jetbrains.idea.maven.importing.MavenImporter
import org.jetbrains.idea.maven.importing.MavenRootModelAdapter
import org.jetbrains.idea.maven.project.MavenProject
import org.jetbrains.idea.maven.project.MavenProjectChanges
import org.jetbrains.idea.maven.project.MavenProjectsProcessorTask
import org.jetbrains.idea.maven.project.MavenProjectsTree
import com.luneo7.junit4.intellij.state.ListenersConfig

class Importer : MavenImporter("org.apache.maven.plugins", "maven-surefire-plugin") {

    override fun preProcess(
        module: Module?,
        mavenProject: MavenProject?,
        changes: MavenProjectChanges?,
        modifiableModelsProvider: IdeModifiableModelsProvider?
    ) {
    }

    override fun process(
        modifiableModelsProvider: IdeModifiableModelsProvider?,
        module: Module?,
        rootModel: MavenRootModelAdapter?,
        mavenModel: MavenProjectsTree?,
        mavenProject: MavenProject?,
        changes: MavenProjectChanges?,
        mavenProjectToModuleName: MutableMap<MavenProject, String>?,
        postTasks: MutableList<MavenProjectsProcessorTask>?
    ) {
        if (module != null) {
            val service = module.getService(ListenersConfig::class.java)

            if (mavenProject != null) {
                val surefire = mavenProject.plugins
                    .first { m -> "maven-surefire-plugin" == m.artifactId }

                if (surefire != null &&
                    surefire.configurationElement != null &&
                    surefire.configurationElement!!.children != null
                ) {

                    val listener = surefire.configurationElement!!
                        .children
                        .filter { e -> e.name == "properties" }
                        .flatMap { e -> e.children }
                        .filter { e -> e.children.any { e1 -> e1.name == "name" && e1.value == "listener" } }
                        .flatMap { e -> e.children.filter { e1 -> e1.name == "value" }.map { e1 -> e1.value } }
                        .first()

                    if (listener != null) {
                        service.state.listeners = listener.split(",").map { e -> e.trim() }
                        return
                    }
                }
            }
            service.state.listeners = emptyList()
        }
    }
}